kafkaesk/app.py

```diff
@@ -236,22 +250,35 @@ async def __run(self) -> None:
                         partition=record.partition,
                         group_id=self._subscription.group,
                     ).inc()
-                except UnhandledMessage:
-                    # how should we handle this? Right now, fail hard
+                except UnhandledMessage as err:
                     logger.warning(f"Could not process msg: {record}", exc_info=True)
-                    # We increase the error counter
                     CONSUMED_MESSAGES.labels(
                         stream_id=record.topic,
                         partition=record.partition,
                         error="UnhandledMessage",
                         group_id=self._subscription.group,
                     ).inc()
+                    try:
+                        await retry_policy(record=record, error=err)
+                    except UnhandledMessage:
+                        # Do not allow this exception up the stack if it is re-raised
+                        ...
+                except Exception as err:
+                    CONSUMED_MESSAGES.labels(
+                        stream_id=record.topic,
+                        partition=record.partition,
+                        error=err.__class__.__name__,
+                        group_id=self._subscription.group,
+                    ).inc()
+                    await retry_policy(record=record, error=err)
```

Comment:
if this retry is encountered, does that mean a message could potentially be retried twice in rapid succession (if the exception is raised from the `retry_policy` call on L262)?

could this have bad side-effects in some cases, e.g. if Kafka or some other external resource is overloaded or unavailable?

also, is there an exception handler for this 2nd retry call, or will that cause the consumer to crash?

Reply:
I'll have to double-check, but I think a lot of this logic is going to move/change with the updates suggested by @vangheem below.

Reply:
gotcha. feel free to resolve this one then and I'll take another look after that.